pkg/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,89 @@
+package garbagecollector
+
+import (
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	winformers "github.com/sdminonne/workflow-controller/pkg/client/informers/externalversions"
+	wlisters "github.com/sdminonne/workflow-controller/pkg/client/listers/workflow/v1"
+	wclientset "github.com/sdminonne/workflow-controller/pkg/client/versioned"
+
+	"github.com/sdminonne/workflow-controller/pkg/controller"
+)
+
+const (
+	// Interval represent the interval to run Garabge Collection
+	Interval time.Duration = 10 * time.Second
+)
+
+// GarbageCollector represents a Workflow Garbage Collector.
+// It collects orphaned Jobs
+type GarbageCollector struct {
+	KubeClient     clientset.Interface
+	WorkflowClient wclientset.Interface
+	WorkflowLister wlisters.WorkflowLister
+	WorkflowSynced cache.InformerSynced
+}
+
+// NewGarbageCollector builds initializes and returns a GarbageCollector
+func NewGarbageCollector(workflowClient wclientset.Interface, kubeClient clientset.Interface, workflowInformerFactory winformers.SharedInformerFactory) *GarbageCollector {
+	return &GarbageCollector{
+		KubeClient:     kubeClient,
+		WorkflowClient: workflowClient,
+		WorkflowLister: workflowInformerFactory.Workflow().V1().Workflows().Lister(),
+		WorkflowSynced: workflowInformerFactory.Workflow().V1().Workflows().Informer().HasSynced,
+	}
+}
+
+// CollectWorkflowJobs collect the orphaned jobs. First looking in the workflow informer list
+// then retrieve from the API and in case NotFound then remove via DeleteCollection primitive
+func (c *GarbageCollector) CollectWorkflowJobs() {
+	glog.V(4).Infof("Collecting garbage jobs")
+	jobs, err := c.KubeClient.BatchV1().Jobs(metav1.NamespaceAll).List(metav1.ListOptions{
+		LabelSelector: controller.WorkflowLabelKey,
+	})
+	if err != nil {
+		glog.Errorf("Garbage collector: unable to list workflow jobs to be collected: %v", err)
+	}
+	collected := make(map[string]struct{})
+	for _, job := range jobs.Items {
+		workflowName, found := job.Labels[controller.WorkflowLabelKey]
+		if !found {
+			glog.Errorf("unable to find workflow name for job: %s/%s", job.Namespace, job.Name)
+			continue
+		}
+		if _, done := collected[path.Join(job.Namespace, workflowName)]; done {
+			continue // already collected so skip
+		}
+		if _, err := c.WorkflowLister.Workflows(job.Namespace).Get(workflowName); err == nil || !apierrors.IsNotFound(err) {
+			if err != nil {
+				glog.Errorf("unable to retrieve workflow %s/%s cache: %v", job.Namespace, workflowName, err)
```

Comment:
A comment like "Unexpected error while using the workflow cache for %s/%s: %v"
Because the IsNotFound error would also be a valid error for your comment...

To make things event more readable I would code it that way:
_, err := c.WorkflowLister.Workflows(job.Namespace).Get(workflowName);
if err== nil {
continue // the workflow is still valid
}
if !apierrors.IsNotFound(err) {
glog.Errorf("Unexpected error while using the workflow cache for %s/%s: %v")
continue
}


Even better maybe with a switch on err, but it will start to be a boomrang code.

Reply:
The idea here is that we look for jobs without a workflow. So we're interested in case where there's error AND the error IsNotFound for DeMorgan law it's what we test in-line... Finally it's the same :)

But it's generally true for the comment. Going to change.  Thanks